File: LinkedList.swift
class LinkedList<T> {
    let data: T
    var next: LinkedList? = nil

    init(data: T) {
        self.data = data
    }

    func append(data: T) {
        if let next = next {
            next.append(data: data)
        } else {
            next = LinkedList(data: data)
        }
    }

    func printList() {
        var currentNode = self
        while let nextNode = currentNode.next {
            print(currentNode.data)
            currentNode = nextNode
        }
        print(currentNode.data)
    }
}

func yieldPairs<T>(
    list: LinkedList<T>,
    yieldFn: (T, T) -> Void
) {
    var a = list
    var b = list.next
    while let bVal = b {
        yieldFn(a.data, bVal.data)
        a = a.next!
        b = bVal.next
    }
}

func appendSorted<T: Comparable>(
    list: LinkedList<T>?,
    data: T
) -> LinkedList<T> {
    if let oldHead = list {
        let newNode = LinkedList(data: data)

        if oldHead.data >= data {
            newNode.next = oldHead
            return newNode
        } else {
            var current = oldHead
            while let next = current.next {
                if next.data >= data {
                    break
                }
                current = next
            }

            newNode.next = current.next
            current.next = newNode

            return oldHead
        }
    } else {
        return LinkedList(data: data)
    }
}

func main() {
    var list = LinkedList(data: 1)
    list = appendSorted(list: list, data: 5)
    list = appendSorted(list: list, data: 2)
    list = appendSorted(list: list, data: 3)
    list = appendSorted(list: list, data: 4)
    list.printList()
    yieldPairs(list: list) { a, b in
        print("\(a), \(b)")
    }
}

main()

File: LinkedList.rs
use std::fmt::Display;

struct LinkedList<T> {
    data: T,
    next: Option<Box<LinkedList<T>>>,
}

impl<T> LinkedList<T> {
    fn new(data: T) -> LinkedList<T> {
        LinkedList {
            data: data,
            next: None,
        }
    }

    fn append(&mut self, data: T) {
        if let Some(ref mut next) = self.next {
            next.append(data);
        } else {
            self.next = Some(Box::new(LinkedList::new(data)));
        }
    }
}

impl<T> LinkedList<T> where T: Display {
    fn print(&self) {
        let mut current = self;
        while let Some(ref next) = current.next {
            println!("{}", current.data);
            current = next;
        }
        println!("{}", current.data);
    }
}

fn yield_pairs<'a, T, F>(
    list: &'a LinkedList<T>,
    yield_fn: F
) where F: Fn(&'a T, &'a T) {
    let mut a = list;
    let mut b = &list.next;
    while let Some(b_val) = b {
        yield_fn(&a.data, &b_val.data);
        a = a.next.as_ref().unwrap();
        b = &b_val.next;
    }
}

fn append_sorted<T: Ord>(
    list: Option<LinkedList<T>>,
    data: T,
) -> LinkedList<T> {
    if let Some(mut old_head) = list {
        let mut new_node = LinkedList::new(data);

        if old_head.data >= new_node.data {
            new_node.next.replace(Box::new(old_head));
            return new_node;
        } else {
            let mut current = &mut old_head;
            while let Some(ref next) = current.next {
                if next.data >= new_node.data {
                    break;
                }
                current = current.next.as_mut().unwrap();
            }

            new_node.next = current.next.take();
            current.next.replace(Box::new(new_node));

            return old_head;
        }
    } else {
        LinkedList::new(data)
    }
}


fn main() {
    let mut list = LinkedList::new(1);
    list = append_sorted(Some(list), 5);
    list = append_sorted(Some(list), 2);
    list = append_sorted(Some(list), 4);
    list = append_sorted(Some(list), 3);
    list.print();
    yield_pairs(&list, |a, b| println!("{}, {}", a, b));
}
